MudDebouncedInput: Fix swallowed user input on component re-render (#7193)

igotinfected opened the pull request:
## Description

Follows #7159 in fixing swallowed user input.

After updating to `6.7.0` we noticed we still had issues with text disappearing on some of our debounced inputs which made me realise the issue was component refreshes in general.

This PR makes it so that `MudDebouncedInput`s don't update the text property unless the caller requests a `Value` update, or the `Value` property was actually changed.

I checked my use case as well as one edge case I thought of, having `Clearable` set to `true` and attempting to clear a value while there is uncommitted text, which works fine.

Let me know if you see any potential issues that you want me to write tests for, or if you have any ideas for a "safer" way to implement this.

Fixes: #7193

## How Has This Been Tested?
- Visually
- Current tests still pass
- Added a new component + accompanying unit test for the described issue, similar to #7159

### Before

https://github.com/MudBlazor/MudBlazor/assets/15004223/0a94ff6c-5c18-45c8-b680-bcd479419165

### After

https://github.com/MudBlazor/MudBlazor/assets/15004223/c7fd18fd-40e8-4adf-b49f-2be2dd3eb761


## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:
- [x] The PR is submitted to the correct branch (`dev`).
- [x] My code follows the code style of this project.
- [x] I've added relevant tests.

igotinfected committed:
MudDebouncedInput: Fix swalled user input on component re-render (#7193)

Fixes: #7193

igotinfected committed:
Fix formatting

henon commented:
Did you test this with WASM or with BSS? Usually such issues behave differently on these architectures, so it is important to make sure they are tested on both.

igotinfected commented:
@henon WASM only. Just gave it a spin and seems like this wasn't an issue on BSS to begin with, due to `TextUpdateSuppression` from `MudBaseInput`:

![image](https://github.com/MudBlazor/MudBlazor/assets/15004223/c7c7924b-a295-4e1b-9e7a-02f74094bdc7)

I tested my test component within the `MudBlazor.Docs.Server` project, including the `Clearable` variation, which (now) behaves identical to the `WASM` version.

I'll run some more thorough tests tomorrow (including on `MudNumericField`) just to be sure though!

igotinfected committed:
Suppress text update only when debounce is active

Fixes a regression where a default `Value` for a debounced field would
not set the `Text` property on initial render, even though the `Value`
was successfully updated.

igotinfected committed:
Add more tests

igotinfected commented:
@henon FYI:

- added more `pre` fix failing tests (`MudTextField` & `MudNumericField`)
![image](https://github.com/MudBlazor/MudBlazor/assets/15004223/c951a60b-81f4-4012-b14b-e6a2030fd3b1)
- fixed a regression that spawned from my changes (default values provided to a component would not apply to the `Text` property and hence not be rendered in the input)
- added tests to cover that regression

As for *BSS*, this is the behaviour I observed:

1. `pre` fix with `TextUpdateSuppression="true"`, does not swallow user input on re-render, but does not update text after debounce (i.e. to apply a new `Culture` or `Format` if they change while the user is typing)
2. `pre` fix with `TextUpdateSuppresion="false"`, swallows user input on re-render, behaves exactly as *WASM* `pre` fix as far as I can tell
3. `post` fix with `TextUpdateSuppression="true"`, as far as I can tell, exactly the same behaviour as `1.`, i.e. no text update after debounce to apply new `Culture`, `Format`, or other
4. `post` fix with `TextUpdateSuppression="false"`, as far as I can tell, exactly the same behaviour as *WASM* `post` fix, i.e. *does* update text after debounce to apply `Culture`, ...

I have a feeling `TextUpdateSuppression` and this PR are fixes for similar issues in BSS and WASM but I'm not sure that this PR covers all of the issues `TextUpdateSuppression` aims to fix. We don't really work with BSS at work currently so I have limited experience with it, and essentially none using MudBlazor.

Let me know if there's anything you want me to mess around with further or if there are any other potential issues you (anyone reading this) can come up with!

henon commented:
TextUpdateSuppression should fix a problem on BSS where writing really fast would lead to swallowed characters for undebounced text fields. The problem with BSS is that there is a lag between the user typing and the server processing it that can cause swallowed characters. This essentially also makes our input mask implementation unreliable in BSS.

henon commented:
I love that you were able to write tests for this issue because I never managed to write any tests for TextUpdateSuppression, it simply didn't happen in bUnit. So I guess the issue may be similar but is also quite different in a way.

henon commented:
Thank you very much @igotinfected , clearly a lot of effort went into this PR.
